File: frontend/src/lib/readonly.ts
import { isHandheldNow } from "./device";

// The web app is an editor on a computer and a viewer on a handheld.
//
// This is the GitHub arrangement: browse anywhere, author where you have the
// keyboard and pointer to do it with. The line is drawn by DEVICE, not by
// window width -- a laptop dragged narrow is still a laptop, and taking its
// editor away because of a window size was simply a bug. See lib/device.ts
// for how a handheld is identified and why no single browser API suffices.

// Named for what the user is trying to do, not for the endpoint behind it, so
// a call site reads as intent: `can("workflow.deploy", readOnly)`.
export type Capability =
  // Authoring — needs the room a desktop has.
  | "workflow.create"
  | "workflow.delete"
  | "workflow.editGraph"
  | "workflow.deploy"
  | "workflow.buildFromChat"
  // Operating — available on any screen.
  | "workflow.run"
  | "workflow.stop"
  | "workflow.chat"
  | "account.billing"
  | "account.settings";

// Everything absent from this set stays available. Listing what is *withheld*
// (rather than what is permitted) means a capability added later is readable
// by default, and has to be denied on purpose.
const WITHHELD: ReadonlySet<Capability> = new Set<Capability>([
  "workflow.create",
  "workflow.delete",
  "workflow.editGraph",
  "workflow.deploy",
  // Build mode rewrites the graph from a chat message, so it is authoring
  // wearing a conversation's clothes. Ordinary chat with an already-deployed
  // workflow is not, and stays open.
  "workflow.buildFromChat",
]);

// Pure on purpose: `readOnly` is passed in rather than measured here, so the
// policy can be tested without a DOM and so React components drive it from
// useReadOnly() -- a function that measured the viewport itself would not
// re-render anything when the window crossed the breakpoint.
export function can(capability: Capability, readOnly: boolean): boolean {
  if (!readOnly) return true;
  return !WITHHELD.has(capability);
}

// Defence in depth for the API layer. Hiding a control is a UX decision; this
// is the guarantee that a missed control, a stale bundle, or a deep link
// cannot still put a write on the wire from a viewer.
//
// The rules deliberately mirror backend/internal/api/readonly.go one for one.
const WRITE_RULES: ReadonlyArray<{ method: string; pattern: RegExp }> = [
  { method: "POST", pattern: /^\/workflows$/ },
  { method: "PUT", pattern: /^\/workflows\/[^/]+$/ },
  { method: "DELETE", pattern: /^\/workflows\/[^/]+$/ },
  { method: "POST", pattern: /^\/workflows\/[^/]+\/deploy$/ },
  { method: "POST", pattern: /^\/workflows\/[^/]+\/build$/ },
  { method: "PUT", pattern: /^\/workflows\/[^/]+\/schedule$/ },
  { method: "DELETE", pattern: /^\/workflows\/[^/]+\/schedule$/ },
  { method: "PUT", pattern: /^\/workflows\/[^/]+\/geofence$/ },
  { method: "DELETE", pattern: /^\/workflows\/[^/]+\/geofence$/ },
  { method: "GET", pattern: /^\/tendril\/console$/ },
  // Same exception, same reason: a GET that find-or-creates the console's
  // workflow row. /prism/console/exists never creates one, and /prism/run is
  // operating rather than authoring, so neither is listed -- exactly the line
  // /tendril/console/exists and /tendril/run already sit on.
  { method: "GET", pattern: /^\/prism\/console$/ },
];

// `path` is the API path as written at the call site (leading slash, no
// origin, no query). Query strings and trailing slashes are normalised off
// first so a caller cannot slip a write past by appending either.
export function isWriteBlocked(
  method: string,
  path: string,
  readOnly: boolean,
): boolean {
  if (!readOnly) return false;
  const m = method.toUpperCase();
  let p = path.split("?")[0].split("#")[0];
  if (p.length > 1 && p.endsWith("/")) p = p.slice(0, -1);
  return WRITE_RULES.some((r) => r.method === m && r.pattern.test(p));
}

// For callers outside React, which is really just the fetch guards below.
// A one-shot reading is the right shape there: it answers "may this call go
// out, right now", and nothing needs to re-render when the answer changes.
// Components must use useReadOnly() instead.
export function isReadOnlyNow(): boolean {
  return isHandheldNow();
}

// Defence in depth for the API layer, shared by every module that calls this
// backend directly from the WEB bundle (lib/api.ts, lib/tendril.ts). Not for
// native/api.ts: that file runs only inside the native Android shell, which
// isHandheldNow() classifies as handheld unconditionally by design (see
// lib/device.ts) -- gating it on this same check would block every native
// write permanently rather than just a web viewer's.
export function assertWritable(method: string, path: string): void {
  if (isWriteBlocked(method, path, isReadOnlyNow())) {
    throw new Error(
      "Workflows can only be edited in the AgentMesh desktop app.",
    );
  }
}

File: backend/internal/api/readonly.go
package api

import (
	"crypto/subtle"
	"log"
	"net/http"
	"os"
	"regexp"
	"strings"

	"github.com/agentmesh/backend/internal/respond"
)

// Read-only mode: refuse to let a client author a workflow.
//
// Note what this is NOT. The frontend decides who may author by viewport --
// a desktop is an editor, a phone is a viewer (frontend/src/lib/readonly.ts) --
// and a server cannot see a viewport. So this middleware cannot enforce that
// split, and does not try to. It answers a coarser question: may ANY client
// write to this deployment at all?
//
// That makes it useful for a deployment that is meant to be wholly read-only --
// a public demo, a mirror, a read replica -- and useless for the ordinary one,
// which is why it is off unless switched on. The client-side split remains a
// UX decision, honestly labelled as such rather than dressed up as a security
// boundary.
//
// Configured from the environment rather than from the request's identity:
// recording *which* clients may write would mean a user- or session-level
// capability, and that means a schema change. Nothing here reads or writes
// the database.

const (
	// Off unless explicitly enabled, so an existing deployment keeps behaving
	// exactly as it does today until someone opts in.
	readOnlyEnvVar = "WEB_READONLY_MODE"

	// The desktop app's way through. It is a shared secret, so it is exactly
	// as strong as that app's ability to keep one -- which is why the frontend
	// never sends it and never learns it.
	editorKeyEnvVar    = "EDITOR_CLIENT_KEY"
	editorKeyHeaderKey = "X-AgentMesh-Editor-Key"
)

// The graph-mutating endpoints, and only those. Running and stopping a
// workflow, reading anything, and every billing or account call are all
// deliberately absent -- a viewer can still operate a workflow somebody else
// built, which is the whole point of the second screen.
//
// Kept in step with WRITE_RULES in frontend/src/lib/readonly.ts: a call this
// list rejects but the frontend permits becomes a control that fails at the
// server with no explanation.
var readOnlyBlocked = []struct {
	method string
	path   *regexp.Regexp
}{
	{http.MethodPost, regexp.MustCompile(`^/workflows$`)},
	{http.MethodPut, regexp.MustCompile(`^/workflows/[^/]+$`)},
	{http.MethodDelete, regexp.MustCompile(`^/workflows/[^/]+$`)},
	{http.MethodPost, regexp.MustCompile(`^/workflows/[^/]+/deploy$`)},
	{http.MethodPost, regexp.MustCompile(`^/workflows/[^/]+/build$`)},
	{http.MethodPut, regexp.MustCompile(`^/workflows/[^/]+/schedule$`)},
	{http.MethodDelete, regexp.MustCompile(`^/workflows/[^/]+/schedule$`)},
	{http.MethodPut, regexp.MustCompile(`^/workflows/[^/]+/geofence$`)},
	{http.MethodDelete, regexp.MustCompile(`^/workflows/[^/]+/geofence$`)},
	// GET, but find-or-creates a workflow row server-side, so it belongs on
	// this list despite the general rule that reads are exempt.
	{http.MethodGet, regexp.MustCompile(`^/tendril/console$`)},
	// Same exception, same reason: find-or-creates the console's workflow
	// row. /prism/console/exists is the non-creating variant and is exempt.
	{http.MethodGet, regexp.MustCompile(`^/prism/console$`)},
}

// blocksWrite reports whether read-only mode rejects this method and path.
// Split out from the middleware so the rules can be tested without standing up
// a router or a request chain.
//
// The trailing slash is normalised away first. To be precise about why, since
// an earlier version of this comment got it wrong: chi does NOT currently
// route "/workflows/" to the "/workflows" handler. With no StripSlashes or
// RedirectSlashes registered in router.go, the trailing-slash form matches no
// route at all and chi answers 404 -- before this middleware is consulted,
// because a group-scoped middleware only runs for paths the router matched.
// TestTrailingSlashNeverReachesTheGroup in router_test.go pins that.
//
// So the normalisation is defence in depth, not a live requirement: the day
// somebody adds slash-handling middleware, these rules must still hold, and
// four lines that fail safe are cheaper than the bypass they would leave.
func blocksWrite(method, path string) bool {
	if len(path) > 1 {
		path = strings.TrimRight(path, "/")
		if path == "" {
			path = "/"
		}
	}
	for _, rule := range readOnlyBlocked {
		if rule.method == method && rule.path.MatchString(path) {
			return true
		}
	}
	return false
}

// NewReadOnlyMiddleware rejects graph-mutating requests with 403 while
// WEB_READONLY_MODE is set. When it is not, the returned middleware is a
// pass-through.
//
// The environment is read once, when the router is built, so a request cannot
// change the mode it is judged under.
func NewReadOnlyMiddleware() func(http.Handler) http.Handler {
	enabled := isTruthy(os.Getenv(readOnlyEnvVar))
	editorKey := os.Getenv(editorKeyEnvVar)

	// Logged because the failure mode is silent: a mistyped variable name
	// leaves the mode OFF and nothing anywhere says so. An operator who set
	// it and reads "disabled" here knows immediately that it did not take.
	// (A typed config loader would be the real fix, but this repo has no
	// config package -- auth.go and middleware.go read the environment the
	// same way -- and introducing one is well outside this change.)
	state := "disabled"
	if enabled {
		state = "enabled"
	}
	log.Printf("web read-only: %s (%s)", state, readOnlyEnvVar)

	return func(next http.Handler) http.Handler {
		if !enabled {
			return next
		}
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if !blocksWrite(r.Method, r.URL.Path) {
				next.ServeHTTP(w, r)
				return
			}
			// An empty EDITOR_CLIENT_KEY must never mean "everyone is an
			// editor", so the bypass is only available once a key is set.
			if editorKey != "" && constantTimeEqual(r.Header.Get(editorKeyHeaderKey), editorKey) {
				next.ServeHTTP(w, r)
				return
			}
			respond.Error(w, http.StatusForbidden,
				"workflows are read-only here; edit them in the AgentMesh desktop app")
		})
	}
}

func isTruthy(v string) bool {
	switch strings.ToLower(strings.TrimSpace(v)) {
	case "1", "true", "yes", "on":
		return true
	}
	return false
}

// Compared in constant time so a wrong key cannot be discovered a byte at a
// time from response timing.
func constantTimeEqual(got, want string) bool {
	return subtle.ConstantTimeCompare([]byte(got), []byte(want)) == 1
}
